WikiWatershed/model-my-watershed
# Mapshed: Split HUC-08 and HUC-10 Into HUC-12 Jobs

## Overview

This PR sets us up for sub-basin modeling: running mapshed/gwlfe on huc8s and huc10s by splitting them into their component huc12s. Later we'll be sending the per-huc12 results to an external API to get the summed result (for the whole AOI) and catchment level water quality results for each huc12.

- Reintroduce feature flagging via environment variables and app settings as decided [here](https://github.com/WikiWatershed/model-my-watershed/blob/develop/doc/arch/adr-001-feature-flags.md)
- Give /mapshed and /gwlfe endpoints a `subbasin` query param
  to toggle if the submitted shape should be subbasin modeled
- Set the frontend up to subbasin model shapes if their applicable and the feature is on (we can't display these results correctly until the external SRAT catchment API is in place)

Connects #2520 

### Demo
<details>

<summary>
Mapshed and GWLF-E Example Results
</summary>


```js
// /mapshed?subbasin=true
// subbasin mapshed results have the huc12 ids as keys
{
  "030300030601": {
    "n38": 0,
    "n39": 0,
    "n30": 0,
    "n31": 0,
    "n32": 0,
    "n33": 0,
    "n34": 0,
    "n35": 0,
    "n36": 0,
    "n37": 0,
    "FCOrgsPerDay": [
      100000000000,
      100000000000,
      140000000,
      140000000,
      11000000000,
      12000000000,
      420000000,
      95000000,
      0
    ],
    "SedNitr": 857.2554744525547,
    ...
   },
  ...
}


// /gwlfe?subbasin=true

{
  "050701020306": {
    "MeanFlow": 30777310.455871146,
    "monthly": [
      {
        "AvWithdrawal": 0,
        "AvEvapoTrans": 0.6061291541535201,
        "AvTileDrain": 0,
        "AvPrecipitation": 7.168726666666667,
        "AvRunoff": 0.668697299714013,
        "AvGroundWater": 4.44275890496891,
        "AvPtSrcFlow": 0,
        "AvStreamFlow": 5.1114562046829235
      },
      {
        "AvWithdrawal": 0,
        "AvEvapoTrans": 0.8965523294793736,
        "AvTileDrain": 0,
        "AvPrecipitation": 7.3566866666666675,
        "AvRunoff": 0.9472849230828789,
         ...
     }, ...
    ], ...
  }, ....
}

```

</details>

### Performance
With my VM set to use...
- 2 CPUS
- 2 celery threads
- timed using `started` and `finished` on job results

#### HUC-10 (`huc10__10017` — Lower Deep River)
**Subbasin**
Mapshed: `35s and 39s` uncached, `8s` cached
GWLF-E: `27s and 36s`

**Develop**
Mapshed: `16s` uncached, `6s` cached
GWLF-E: `6s`

#### HUC-08 (`huc8__63` — Elk)
**Subbasin**
Mapshed: `40s and 69s` (ran twice uncached), `26s` cached
GWLF-E: `30s and 56s`


**Develop**
Mapshed: `38s` uncached, `8s` cached
GWLF-E: `6s`

## Notes
- ~Originally when I tried to run HUC-08s, I'd lose connection to one of my celery threads. I added a countdown to space out the task submission in ed9505d, and haven't seen the issue since. Currently testing if it indeed fixed it, or if the issue earlier was just a fluke.~ I've determined this was a fluke and deleted the commit that adds the countdown.

- Watersheds with many huc12s like Lower Savannah (19 of them) result in a mapshed result of `5.516278MB`, which is greater than our 5MB request body limit. We could increase this limit or work around having to send the GMS results back. For example, we could send the job id to `/gwlfe` and have the server look up the result.

## Testing Instructions

 * `vagrant provision app` and `./scripts/toggle_feature.sh subbasin`
 * `./scripts/bundle.sh` and confirm `clientSettings.enabled_feautures = ["subbasin"]`
 * Select huc12s and other non huc8/huc10 shapes. They should complete mapshed and gwlfe as before
 * Select a huc10 and confirm the mapshed and gwlfe results are in the form of a dictionary of huc12s (the charts in the UI will be blank)
 * Note the results of one of the huc12's and its watershed id 
  * Run mapshed on the huc12 and confirm its results are the same as that returned in the subbasin dictionary result
* Confirm you can also subbasin model huc8s

**If you need to clear your cache to play with performance uncached**
```sh
vagrant ssh app
envdir /etc/mmw.d/env bash -c 'redis-cli -h $MMW_CACHE_HOST -n 1 --raw KEYS ":1:geop*" | xargs redis-cli -h $MMW_CACHE_HOST -n 1 DEL'
```
